modules/livestore/live_store.go

```diff
@@ -0,0 +1,421 @@
+	}
+
+	partitionRingKV := cfg.PartitionRing.KVStore.Mock
+	if partitionRingKV == nil {
+		partitionRingKV, err = kv.NewClient(cfg.PartitionRing.KVStore, ring.GetPartitionRingCodec(), kv.RegistererWithKVName(reg, ingester.PartitionRingName+"-lifecycler"), logger)
+		if err != nil {
+			return nil, fmt.Errorf("creating KV store for ingester partition ring: %w", err)
+		}
+	}
+
+	s.ingestPartitionLifecycler = ring.NewPartitionInstanceLifecycler(
+		s.cfg.PartitionRing.ToLifecyclerConfig(s.ingestPartitionID, cfg.LifecyclerConfig.ID),
+		ingester.PartitionRingName,
+		ingester.PartitionRingKey,
+		partitionRingKV,
+		logger,
+		prometheus.WrapRegistererWithPrefix("tempo_", reg))
+
+	s.subservicesWatcher = services.NewFailureWatcher()
+	s.subservicesWatcher.WatchService(s.ingestPartitionLifecycler)
+
+	s.Service = services.NewBasicService(s.starting, s.running, s.stopping)
+
+	return s, nil
+}
+
+func (s *LiveStore) starting(ctx context.Context) error {
+	var err error
+	s.wal, err = wal.New(&s.cfg.WAL)
+	if err != nil {
+		return fmt.Errorf("failed to create WAL: %w", err)
+	}
+
+	err = services.StartAndAwaitRunning(ctx, s.ingestPartitionLifecycler)
+	if err != nil {
+		return fmt.Errorf("failed to start partition lifecycler: %w", err)
+	}
+
+	s.client, err = ingest.NewReaderClient(
+		s.cfg.IngestConfig.Kafka,
+		ingest.NewReaderClientMetrics(liveStoreServiceName, s.reg),
+		s.logger,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create kafka reader client: %w", err)
+	}
+
+	boff := backoff.New(ctx, backoff.Config{
+		MinBackoff: 100 * time.Millisecond,
+		MaxBackoff: time.Minute, // If there is a network hiccup, we prefer to wait longer retrying, than fail the service.
+		MaxRetries: 10,
+	})
+
+	for boff.Ongoing() {
+		err := s.client.Ping(ctx)
+		if err == nil {
+			break
+		}
+		level.Warn(s.logger).Log("msg", "ping kafka; will retry", "err", err)
+		boff.Wait()
+	}
+	if err := boff.ErrCause(); err != nil {
+		return fmt.Errorf("failed to ping kafka: %w", err)
+	}
+
+	s.reader, err = NewPartitionReaderForPusher(s.client, s.ingestPartitionID, s.cfg.IngestConfig.Kafka, s.consume, s.logger, s.reg)
+	if err != nil {
+		return fmt.Errorf("failed to create partition reader: %w", err)
+	}
+	err = services.StartAndAwaitRunning(ctx, s.reader)
+	if err != nil {
+		return fmt.Errorf("failed to start partition reader: %w", err)
+	}
+
+	s.wg.Add(1)
+	go s.completeLoop()
+
+	return nil
+}
+
+func (s *LiveStore) running(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return nil
+	case err := <-s.subservicesWatcher.Chan():
+		return fmt.Errorf("live-store subservice failed: %w", err)
+	}
+}
+
+func (s *LiveStore) stopping(error) error {
+	// Stop consuming
+	err := services.StopAndAwaitTerminated(context.Background(), s.reader)
+	if err != nil {
+		level.Warn(s.logger).Log("msg", "failed to stop reader", "err", err)
+		return err
+	}
+
+	// Cancel and wait for async loops to return
+	s.cancel()
+	s.flushqueues.Close()
```

Comment:
Should this be `DiscardAndClose`? The `Dequeue` call in the `completeLoop` looks like it would still try consuming all the items from the queue which could ultimately exceed the grace period shutdown and prevent a clean teardown.